runtime/common/ArgumentConversion.cpp

```diff
@@ -17,6 +17,34 @@
 
 using namespace mlir;
 
+cudaq::opt::StateData
+cudaq::opt::StateData::readStateData(const cudaq::state *state) {
+  auto precision = state->get_precision();
+  auto stateVector = state->get_tensor();
+  auto numElements = stateVector.get_num_elements();
+  auto elementSize = stateVector.element_size();
+
+  if (state->is_on_gpu()) {
+    if (precision == cudaq::SimulationState::precision::fp32) {
+      assert(elementSize == sizeof(std::complex<float>) &&
+             "Incorrect complex<float> element size");
+      auto *hostData = new std::complex<float>[numElements];
+      state->to_host(hostData, numElements);
```

Comment:
I'm kind of lost why we need to do a deep copy of the data here.

At this point, we're in the process of launching the kernel. The bits in the `cudaq::state` object are final and cannot change at this point. So why copy them? Instead of copying them, we should be able to do the $n$ dereferences to the actual floating point bits and either construct a copy in MLIR code or skip the last dereference and pass the pointer to the block of memory.

So I don't think we want to pass a reference to the `cudaq::state` object (which is a wrapper around a shared_ptr) or reference to the shared_ptr, but the pointer inside the shared_ptr. And then we have to choice of using that pointer or dereferencing it to read out the data. What is the simulator expecting here? A vector? Which is yet another wrapper class in the way. Or maybe just something like a raw `double*`?

At any rate, I'm not sure we need this StateData object if we don't need to make a copy of the bits or create an object to manage that copy.

Reply:
The code will execute in a different process or on a different machine, so we need to copy and materialize the data so the server code will know what to use. Another option is to replace the quake.init_state by a call to a kernel that generates the data but it has a lot of work to do, and might not be ideal for remote simulation scenario (to be determined)

Reply:
Removed the `StateData` class, replaced by getting data values from the state using `state::operator()` instead.